diff -c /Volumes/Linux/preware/toolchain/ipkg-utils/build/ipkg-utils/arfile.py\~ /Volumes/Linux/preware/toolchain/ipkg-utils/build/ipkg-utils/arfile.py
*** ipkg-utils/arfile.py~	2009-08-12 20:35:30.000000000 +0930
--- ipkg-utils/arfile.py	2009-08-12 21:04:19.000000000 +0930
***************
*** 77,83 ****
              descriptor = l.split()
  #            print descriptor
              size = int(descriptor[5])
!             memberName = descriptor[0][:-1]
              self.directory[memberName] = descriptor + [self.f.tell()]
  #            print "read:", memberName
              if memberName == fname or (memberName.startswith("`") and memberName[1:] == fname):
--- 77,84 ----
              descriptor = l.split()
  #            print descriptor
              size = int(descriptor[5])
!             memberName = descriptor[0]
!             if memberName[-1] == '/': memberName = memberName[:-1]
              self.directory[memberName] = descriptor + [self.f.tell()]
  #            print "read:", memberName
              if memberName == fname or (memberName.startswith("`") and memberName[1:] == fname):
***************
*** 122,126 ****
          tarf = tarfile.open("control.tar.gz", "r", tarStream)
          #tarf.list()
  
!         f2 = tarf.extractfile("control")
          print f2.read()
--- 123,127 ----
          tarf = tarfile.open("control.tar.gz", "r", tarStream)
          #tarf.list()
  
!         f2 = tarf.extractfile("./control")
          print f2.read()

