--- ipkg-utils/ipkg-build.orig	2009-08-13 18:08:43.000000000 +0930
+++ ipkg-utils/ipkg-build	2009-08-13 18:11:08.000000000 +0930
@@ -60,14 +60,15 @@
 	    fi
 	fi
 
-	large_uid_files=`find . -uid +99 || true`
-
-	if [ "$ogargs" = "" ]  && [ -n "$large_uid_files" ]; then
-		echo "*** Warning: The following files have a UID greater than 99.
-You probably want to chown these to a system user: " >&2
-		ls -ld $large_uid_files
-		echo >&2
-	fi
+# Commented out to allow this script to run on MacOSX.
+#	large_uid_files=`find . -uid +99 || true`
+#
+#	if [ "$ogargs" = "" ]  && [ -n "$large_uid_files" ]; then
+#		echo "*** Warning: The following files have a UID greater than 99.
+#You probably want to chown these to a system user: " >&2
+#		ls -ld $large_uid_files
+#		echo >&2
+#	fi
 	    
 
 	if [ ! -f "$CONTROL/control" ]; then
@@ -235,7 +236,7 @@
 pkg_file=$dest_dir/${pkg}_${version}_${arch}.ipk
 rm -f $pkg_file
 if [ "$outer" = "ar" ] ; then
-  ( cd $tmp_dir && ar -crf $pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz )
+  ( cd $tmp_dir && ar -cr $pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz )
 else
   ( cd $tmp_dir && tar -zcf $pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz )
 fi
