--- shellinabox/vt100.js~	2009-11-22 09:17:49.000000000 +1030
+++ shellinabox/vt100.js	2010-12-18 12:58:40.000000000 +1030
@@ -2248,6 +2248,42 @@
   }
 };
 
+VT100.prototype.handleKey = function(event) {
+    /*
+    this.vt100('H: c=' + event.charCode + ', k=' + event.keyCode +
+	       (event.shiftKey || event.ctrlKey || event.altKey ||
+		event.metaKey ? ', ' +
+		(event.shiftKey ? 'S' : '') + (event.ctrlKey ? 'C' : '') +
+		(event.altKey ? 'A' : '') + (event.metaKey ? 'M' : '') : '') +
+	       '\r\n');
+    */
+
+  var ch  = event.charCode;
+  var key = event.keyCode;
+
+  if (ch == 0) {
+      switch (key) {
+      case   8: /* Backspace    */ ch = '\u007f';                       break;
+      case  13: /* Enter        */ ch = this.crLfMode ?
+                                        '\r\n' : '\r';                  break;
+      default:
+	  this.vt100('H: c=' + event.charCode + ', k=' + event.keyCode +
+		     (event.shiftKey || event.ctrlKey || event.altKey ||
+		      event.metaKey ? ', ' +
+		      (event.shiftKey ? 'S' : '') + (event.ctrlKey ? 'C' : '') +
+		      (event.altKey ? 'A' : '') + (event.metaKey ? 'M' : '') : '') +
+		     '\r\n');
+	  return;
+      }
+  }
+
+  if (this.menu.style.visibility == 'hidden') {
+    this.keysPressed(ch);
+  }
+
+  return;
+};
+
 VT100.prototype.inspect = function(o, d) {
   if (d == undefined) {
     d       = 0;
@@ -2451,6 +2486,10 @@
   return true;
 };
 
+VT100.prototype.keyDown = function(event) {
+    return false;
+};
+
 VT100.prototype.keyPressed = function(event) {
   // this.vt100('P: c=' + event.charCode + ', k=' + event.keyCode +
   //            (event.shiftKey || event.ctrlKey || event.altKey ||
@@ -2487,6 +2526,11 @@
   return false;
 };
 
+VT100.prototype.keyPressed = function(event) {
+    this.handleKey(event);
+    return false;
+};
+
 VT100.prototype.keyUp = function(event) {
   // this.vt100('U: c=' + event.charCode + ', k=' + event.keyCode +
   //            (event.shiftKey || event.ctrlKey || event.altKey ||
@@ -2559,6 +2603,17 @@
   return false;
 };
 
+VT100.prototype.keyUp = function(event) {
+
+    switch (event.keyCode) {
+    case   8:
+    case  13:
+	this.handleKey(event); break;
+    }
+
+    return false;
+};
+
 VT100.prototype.animateCursor = function(inactive) {
   if (!this.cursorInterval) {
     this.cursorInterval     = setInterval(
